File: solutions/problem_2347_best_poker_hand/solution.cpp
/*
 * Problem 2347: Best Poker Hand
 * ===========================
 * Difficulty: Easy
 * Tags: Array, Hash Table, Counting
 * Pattern: Hash Map Lookup
 *
 * Time Complexity:  O(n)
 * Space Complexity: O(n)
 */

#include <string>
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    string bestHand(vector<int>& ranks, vector<string>& suits) {
        // Hash map approach - O(n) time, O(n) space
        unordered_map<int, int> seen;
        for (int i = 0; i < ranks.size(); i++) {
            int complement = suits - ranks[i];
            if (seen.count(complement)) {
                return {seen[complement], i};
            }
            seen[ranks[i]] = i;
        }
        return "";
    }
};

File: solutions/problem_2347_best_poker_hand/solution.py
"""
Problem 2347: Best Poker Hand
===========================
Difficulty: Easy
Tags: Array, Hash Table, Counting
Pattern: Hash Map Lookup

Time Complexity:  O(n)
Space Complexity: O(n)
"""

class Solution:
    def bestHand(self, ranks: List[int], suits: List[str]) -> str:
        # Hash map approach - O(n) time, O(n) space
        seen = {}
        for i, val in enumerate(ranks):
            complement = suits - val
            if complement in seen:
                return [seen[complement], i]
            seen[val] = i
        return ""
